crates/db/src/db.rs

```diff
@@ -50,31 +50,52 @@ impl Db {
     &self,
     limit: Option<i64>,
     offset: Option<u64>,
-    course_subjects: Option<Vec<String>>,
-    course_levels: Option<Vec<String>>,
-    course_terms: Option<Vec<String>>,
+    filter: Option<CourseFilter>,
   ) -> Result<Vec<Course>> {
     let mut document = Document::new();
 
-    if let Some(course_subjects) = course_subjects {
-      document.insert(
-        "subject",
-        doc! { "$regex": format!("^({})", course_subjects.join("|")) },
-      );
-    }
+    if let Some(filter) = filter {
+      let CourseFilter {
+        subjects,
+        levels,
+        terms,
+        query,
+        ..
+      } = filter;
+
+      if let Some(subjects) = subjects {
+        document.insert(
+            "subject",
+            doc! { "$regex": format!("^({})", subjects.join("|")), "$options": "i" },
+        );
+      }
 
-    if let Some(course_levels) = course_levels {
-      document.insert(
-        "code",
-        doc! { "$regex": format!("^({})", course_levels.join("|")) },
-      );
-    }
+      if let Some(levels) = levels {
+        document.insert(
+            "code",
+            doc! { "$regex": format!("^({})", levels.join("|")), "$options": "i" },
+        );
+      }
 
-    if let Some(course_terms) = course_terms {
-      document.insert(
-        "terms",
-        doc! { "$regex": format!("^({})", course_terms.join("|")) },
-      );
+      if let Some(terms) = terms {
+        document.insert(
+          "terms",
+          doc! { "$regex": format!("^({})", terms.join("|")), "$options": "i" },
+        );
+      }
+
+      if let Some(query) = query {
+        document.insert(
+            "$or",
+            vec![
+                doc! { "_id": doc! { "$regex": format!(".*{}.*", query.replace(' ', "")), "$options": "i" } },
+                doc! { "title": doc! { "$regex": format!(".*{}.*", query), "$options": "i" } },
+                doc! { "code": doc! { "$regex": format!(".*{}.*", query), "$options": "i" } },
+                doc! { "subject": doc! { "$regex": format!(".*{}.*", query), "$options": "i" } },
+                doc! { "description": doc! { "$regex": format!(".*{}.*", query), "$options": "i" } }
```

Comment:
Why not use the text index like with `text_search` (the thing used for the course search bar)?